### PR TITLE
Feature: flashcard model

### PR DESCRIPTION
**Pull request by Roshan-y**
# Description

This PR is for the flashcard model. I have created the data class for flashcard, the flashcard repository, the flashcard repository Firestore and the flash card view model. The flashcard repository contains the functions getNewUid, init, getFlashcards, getFlashcardById, getFlashcardsByFolder, addFlashcard, updateFlashcard and deleteFlashcard. The view model contains 3 different state flow. flashcards contains the list of flashcards of the user, flashcard is the selected flashcard and selectedFolderFlashcards is the list of flashcards for the current folder.

# How Has This Been Tested?
I have written unit tests for flashcard repository firestore and flashcard view model.

### Unit Tests
The flashcard repository firestore unit tests test if the functions call the correct mock firebase functions and checks if the success and failure callbacks function as intended. The flashcard view model tests test if the view model functions calls the corresponding repository functions.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt
Comment (robertaleG) [inline app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt:115]: if you change the init (see comment) change the test here as well

File: app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt
Comment (robertaleG) [inline app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt:141]: You could replace this asserts by `assertEquals(convertedFlashcard, flashcard)` which tests equality for all fields and shows neatly the differences in the error logs. 

To see if the "==" is correct, add unit tests for the flashcard data class in a new file FlashcardTest.kt. The responsibility of making sure the equals operator works is not the job of FlashcardRepositoryFirestoreTest.kt but FlashcardTest.kt

File: app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt
Comment (robertaleG) [inline app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt:201]: same

File: app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt
Comment (robertaleG) [inline app/src/test/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestoreTest.kt:280]: clean concise code for exception testing, nice job!

**Review by robertaleG — APPROVED** [review]
Some minor changes requested (see comments), overall good code and testing is very thorough. I will approve so you can merge easily after making the changes.

**Comment by robertaleG** [comment]
@all don't merge this yet, wait for Roshan to read my comments before merging

File: app/src/main/java/com/github/onlynotesswent/model/flashcard/FlashcardRepository.kt
Comment (Roshan-y) [reply app/src/main/java/com/github/onlynotesswent/model/flashcard/FlashcardRepository.kt:15]: I was thinking when we call init in the view model for the OnSuccess callback we could call getFlashcardsFrom using the auth user uid (once we modify the user uid to be equal to the user model uid) which in that case we would need auth.

Comment (robertaleG) [reply app/src/main/java/com/github/onlynotesswent/model/flashcard/FlashcardRepository.kt:15]: we can probably do that automatically in overview like I did for the notes (which get updated once the currentUser.value goes from null to non-null)

File: app/src/main/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestore.kt
Comment (robertaleG) [reply app/src/main/java/com/github/onlynotesswent/model/flashcard/FlashcardRepositoryFirestore.kt:70]: fair enough

**Pull request merged**
